designs/nodepool-node-limits.md

```diff
@@ -0,0 +1,144 @@
+# NodePool Node Limits in Karpenter
+
+## Motivation
+
+Service providers and cluster administrators should have the ability to enforce maximum node counts per `NodePool` to address several important use cases:
+
+* **Licensing constraints**: Respect software licensing limits (e.g., monitoring agents with per-node licensing)
+* **IP address management**: With fixed IP address pools, physically prevent node creation beyond the available IP capacity
+* **Service Provider SLAs**: Guarantee service level agreements by preventing unlimited node provisioning
+* **Resource limits**: Respect cloud provider account limits or organizational quotas with Karpenter-level control rather than relying on infrastructure-level restrictions
+
+## Notes
+
+- Enforcing global maximum Node limits is not a goal of this design.
+- This design does not involve nodes managed outside of Karpenter.
+- The Node limits discussed in this design are different from the Node limits discussed in the [Karpenter - Static Capacity](static-capacity.md) design, but involve similar concepts.
+    - See the [Relationship to Static Capacity Design](#relationship-to-static-capacity-design) section for more details.
+
+## Design Options
+
+There are three potential approaches for implementing `NodePool` node limits, each with different trade-offs:
+
+### Option 1: Hard Limits
+
+**Description**: Node limits are enforced as hard constraints that block both provisioning and graceful disruption (Consolidation and Drift) when the limit is reached. Graceful disruption is blocked because Karpenter requires the new nodes to be schedulable and ready before the old node can be terminated, which can lead to more nodes than the limit at one state.
+
+**Benefits**:
+- Guarantees the invariant that node count never exceeds the limit
+- Consistent behavior across all operations
+
+**Drawbacks**:
+- Can block disruption operations when at the limit, which may lead to suboptimal resource utilization
+
+### Option 2: Soft Limits
+
+**Description**: Node limits are enforced only during provisioning, allowing graceful disruption operations to continue even when at the limit.
+
+**Benefits**:
+- Graceful disruption functions continue to work, allowing Consolidation and Drift even at the limit.
+- In steady state, nodes will stay under the limit (assuming the cloud provider allows the node limit to be exceeded)
+
+**Drawbacks**:
+- Intermittent states may temporarily exceed the node limit during disruption operations
```

Comment:
I think this drawback makes this a non-option for the goals listed in the motivation section. It's worth having the proposal since this is the most straightforward extension of Karpenter's existing limits mechanism, but I'd explicitly call out that it doesn't meet the requirements of the RFC.